src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1906,204 @@ MutableBigInteger sqrt() {
         }
     }
 
+    /**
+     * Calculate the integer square root {@code floor(sqrt(this))} and the remainder
+     * if needed, where {@code sqrt(.)} denotes the mathematical square root.
+     * The contents of {@code this} are <b>not</b> changed.
+     * The value of {@code this} is assumed to be non-negative.
+     * @throws ArithmeticException if the value returned by {@code bitLength()}
+     * overflows the range of {@code int}.
+     * @return the integer square root of {@code this} and the remainder if needed
+     */
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
+        // Special cases.
+        if (this.isZero()) {
+            return new MutableBigInteger[] {
+                    new MutableBigInteger(), needRemainder ? new MutableBigInteger() : null
+            };
+        } else if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+
+            if (this.intLen == 1) {
+                if (x < 4) // result is unity
+                    return new MutableBigInteger[] {
+                            ONE, needRemainder ? new MutableBigInteger((int) (x - 1)) : null
+                    };
+
+                long s = (long) Math.sqrt(x); // Math.sqrt is exact
+                return new MutableBigInteger[] {
+                        new MutableBigInteger((int) s),
+                        needRemainder ? new MutableBigInteger((int) (x - s * s)) : null
+                };
+            }
+
+            // Initial estimate is the square root of the unsigned long value.
+            long s = (long) Math.sqrt(x > 0 ? x : x + 0x1p64);
```

Comment:
As a matter of simplification, the above cases can be collapsed to just this one. I don't think there's a real need to optimize for 0, [1, 4), `int`.

Also, there should be a proof that the code below is correct, as there are many roundings involved.

```suggestion
            long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
```

Reply:
The proof has been made simply by exaustive experiment: for every long value `s` in [0, 2^32) such that `x == s * s`, it is true that `s == (long) Math.sqrt(x >= 0 ? x : x + 0x1p64)`. This can be verified directly by a Java program.
It means that Math.sqrt() returns the correct value for every perfect square, and so the value returned by Math.sqrt() for whatever long value in the range [0, 2^64) is either correct, or rounded up by one if the value is too high and too close to a perfect square.

Reply:
(Yes, I proved it to myself in this way.)

A similar explanation should be in a comment in the code, not here. The source of truth is the code, not the comments on GitHub ;-)

Reply:
I would keep zero as the only special case, so as not to have to modify `x > 0` with `x >= 0` in `Math.sqrt(x > 0 ? x : x + 0x1p64)`.

Reply:
I can't see any real-world performance benefit in having a special case for `0` since, functionally, the code for the generic (unsigned) `long` works correctly. Except if there's evidence that a `0` argument is used in a vast majority of cases, which I doubt.
The aim is to simplify as much as reasonably possible without incurring real performance issues.

Changing `>` to `>=` makes the code correct even when nothing is assumed about the range of `x`.